pyuvdata/uvdata/miriad.py

```diff
@@ -1057,10 +1222,23 @@ def write_miriad(
         uv.add_var("pol", "i")
         uv.add_var("lst", "d")
         uv.add_var("cnt", "d")
+        if self.multi_object:
+            uv.add_var("source", "a")
         uv.add_var("ra", "d")
         uv.add_var("dec", "d")
         uv.add_var("inttime", "d")
 
+        app_record = False
+        if (
+            self.phase_center_app_ra is not None
+            and self.phase_center_app_dec is not None
+            and self.phase_center_app_pa is not None
+        ):
+            app_record = True
+            uv.add_var("obsra", "d")
+            uv.add_var("obsdec", "d")
+            uv.add_var("obspa", "d")  # Do we want to add this new keyword?
```

Comment:
It is not a standard MIRIAD keyword, and I'm kind of two minds about this. On the one hand, it doesn't seem like a problem to add extra keywords, but of course, any MIRIAD software won't know what to do with this value (a similar issue exists w/ SMA's MIR format, although we don't have a writer... yet). One alternative would be to undo the position angle fix when writing to MIRIAD, and apply it on reading it (not unlike changing conjugation schemes). I think that would work for CARMA and SMA, although it might be worth asking the PAPER/HERA folks.

Reply:
I think MIRIAD stores parallactic angle in a somewhat roundabout way.

The `chi` keyword specifies "Position angle of the X feed relative to the sky. This is the sum of the parallactic angle and the evector variable."

The `evector` keyword specifies "Position angle of the X feed, to the local vertical."

So the `obspa` value can be computed as `obspa = chi - evector`.  This then gets into issue #854.

Reply:
So `obs_pa ` is getting at something _slightly_ different, and perhaps needs renaming in order to avoid confusion. The PA you are describing above defines the angle between the horizontal coordinate system (i.e., AltAz), and the celestial coordinate system (i.e., ICRS). What `obs_pa` is trying to capture is the difference between apparent coordinates (apparent RA and apparent Dec) and the coordinate system you want to make maps in. It effectively allows you to rotate the map so that "up" points towards the pole in the frame that you are making maps in (e.g., ICRS). 

For most applications, the above effect is relatively small (~1%), but it is baseline-length dependent, and slowly grows with time due to precession, etc.  MIRIAD and MIR seem to neglect this (although MIRIAD does allow for different directions of "up" for mapping planets), but CASA seems to have it accounted for, which I think argues for it being tracked in pyuvdata. But I'm happy to be convinced otherwise.

Reply:
Ah, I see.  That's definitely a different beast.  I guess if one rotates the baseline vector frame from XYZ (or ENU) to UVW such that W points to the local hour angle and declination of `obsra` and `obsdec` at a given time, then I guess U and V will be aligned with "terrestrial" east and north rather than "celestial" east and north, and `obspa` is essentially the difference in rotation around W between the two.  Is that about right?

Reply:
Yup, that's about right.

I was actually thinking about renaming `obspa` as `phase_center_frame_pa` -- that is, the PA of the terrestrial frame at a given apparent RA/Dec against the celestial frame. That would give space for using `obspa` in it's "normal" way, and give a route to tackle #854 (with an additional piece of metadata -- what MIRIAD calls the `evector`)

Reply:
👍 on `phase_center_frame_pa`.  That seems like a much more descriptive term.  Just make sure the direction is well documented (clockwise-vs-anticlockwise and from-vs-to etc.) 😅

Reply:
How is `phase_center_frame_pa` calculated?  Is it by computing the `obsra`/`obsdec` of a slightly offset position and then calculating the rotation compared to the same offset from the original `obsra/pbsdec`?  IOW, is it something like:
```
obsra1, obsdec1 = get_obsra_obsdec(ra, dec, other_params)
onsra2, obsdec2 = get_obsra_obsdec(ra, dec+offset, other_params)
phase_center_frame_pa = atan2(obsdec2-obsdec1, obsra2-obsra1)
```
where `atan2` could be replaced by proper spherical geometry, but is used here for simplicity of the example and small angle approximation is probably OK for small offsets or is there a more analytical approach?

Reply:
That's right - position angle is calculated from a pair of offset positions (~0.5 deg off), using  a similar (although slightly more complicated) formula:
```
obspa = atan2(sin(ra_d - ra_u), cos(dec_d)*tan(dec_u) - sin(dec_d)*cos(ra_u - ra_d)),
```
where `ra_u` and `dec_u` are the coordinates of the "up" position, and `ra_d` and `dec_d` are the coordinates of the "down" position. In testing, this was accurate to about a bit better than one part in 10^6, which seemed sufficient for its use. I toyed around with a couple of other ideas,  including trying a more analytic approach, but they introduced some undesirable corner cases that I couldn't figure out how to get around without adding a lot of complexity. But the `phase_center_frame_pa` is well-contained within a single function, so if someone wanted to swap it out with something better, it would be a relatively straightforward drop & replace.

Reply:
I think we now have two examples showing that your definition of "slightly" is slightly different from mine :)

This looks similar to (if not the same as?) the SOFA/ERFA function `pas()`: https://github.com/liberfa/erfa/blob/master/src/pas.c

Reply:
Presumably `ra_u` et al. are the "observed" positions since otherwise this would always return 0, right?  The offset is in declination only, right?

Reply:
> I think we now have two examples showing that your definition of "slightly" is slightly different from mine :)

I'm slightly amused ;)

> This looks similar to (if not the same as?) the SOFA/ERFA function `pas()`

Yup, looks like its identical -- seems like that's a much better reference than Wikipedia.

> Presumably `ra_u` et al. are the "observed" positions since otherwise this would always return 0, right? The offset is in declination only, right?

That's correct on all accounts above - apologies for not making that more clear. Dec-only was used since you can trust that you're always moving along a great circle on the sphere (not true in RA, of course).

Reply:
Okay, I've renamed the relevant attribute `phase_center_frame_pa` to (hopefully) make things more human-interpretable, as well as swapping in the pyERFA's `pas` function. I've also put into `utils` a function that will calculate the position angle of the feed/antenna relative to the celestial pole, but I have not put in a corresponding attribute into `UVData`, since it's easy enough to calculate on the fly, and it's not clear its needed _yet_. My inclination is to drop this extra keyword from MIRIAD, and just undo/redo the correction when reading/writing from MIRIAD files, since as far as I know, no other MIRIAD-writing facility would have put this correction in.

Reply:
How does one specify the position angle of the X/Y feed arms relative to the antenna's vertical/horizontal?  Interestingly, MeerKAT refers to its linearly polarized feed arms literally as "horizontal" and "vertical" rather than "X" and "Y".  Not sure how they are oriented wrt each other so I guess there's (at least?) a 180 degree unknown (to me) in there.

Reply:
We will need to add an extra keyword to capture what MIRAID typically calls the `evector` value -- I'm certainly amenable to that, although I think that might be better addressed in a separate PR, since there may be other things that need implementation. Hopefully the `calc_pos_angle` function (and the other changes here) will make that work a bit easier :)